server/src/main/java/org/elasticsearch/cluster/coordination/ClusterFormationFailureHelper.java

```diff
@@ -114,17 +120,120 @@ public String toString() {
+            Settings settings,
+            ClusterState clusterState,
+            List<TransportAddress> resolvedAddresses,
+            List<DiscoveryNode> foundPeers,
+            long currentTerm,
+            ElectionStrategy electionStrategy,
+            StatusInfo statusInfo,
+            List<JoinStatus> inFlightJoinStatuses
+        ) {
+            this(
+                INITIAL_MASTER_NODES_SETTING.get(settings),
+                clusterState.nodes().getLocalNode(),
+                clusterState.nodes().getMasterNodes(),
+                clusterState.version(),
+                clusterState.term(),
+                clusterState.getLastAcceptedConfiguration(),
+                clusterState.getLastCommittedConfiguration(),
+                resolvedAddresses,
+                foundPeers,
+                currentTerm,
+                calculateHasDiscoveredQuorum(
+                    foundPeers,
+                    electionStrategy,
+                    clusterState.nodes().getLocalNode(),
+                    currentTerm,
+                    clusterState.term(),
+                    clusterState.version(),
+                    clusterState.getLastCommittedConfiguration(),
+                    clusterState.getLastAcceptedConfiguration()
+                ),
+                statusInfo,
+                inFlightJoinStatuses
+            );
+        }
+
+        private static boolean calculateHasDiscoveredQuorum(
+            List<DiscoveryNode> foundPeers,
+            ElectionStrategy electionStrategy,
+            DiscoveryNode localNode,
+            long currentTerm,
+            long acceptedTerm,
+            long clusterStateVersion,
+            VotingConfiguration lastCommittedConfiguration,
+            VotingConfiguration lastAcceptedConfiguration
+        ) {
+            final VoteCollection voteCollection = new VoteCollection();
+            foundPeers.forEach(voteCollection::addVote);
+            return electionStrategy.isElectionQuorum(
+                localNode,
+                currentTerm,
+                acceptedTerm,
+                clusterStateVersion,
+                lastCommittedConfiguration,
+                lastAcceptedConfiguration,
+                voteCollection
+            );
+        }
+
+        public ClusterFormationState(StreamInput in) throws IOException {
+            this(
+                in.readStringList(),
+                new DiscoveryNode(in),
+                in.readMap(StreamInput::readString, DiscoveryNode::new),
+                in.readLong(),
+                in.readLong(),
+                new VotingConfiguration(in),
+                new VotingConfiguration(in),
+                in.readImmutableList(TransportAddress::new),
+                in.readImmutableList(DiscoveryNode::new),
+                in.readLong(),
+                in.readBoolean(),
+                readStatusInfo(in),
+                in.readList(
+                    in1 -> new JoinStatus(
```

Comment:
could we have a more meaningful name for the inner reader?

Reply:
JoinStatus is now Writeable and readable from the wire.